api/equality/equality.go

```diff
@@ -27,3 +30,50 @@ func TaskStatusesEqualStable(a, b *api.TaskStatus) bool {
 	copyA.Timestamp, copyB.Timestamp = nil, nil
 	return reflect.DeepEqual(&copyA, &copyB)
 }
+
+// RootCAEqualStable compares RootCAs, excluding join tokens, which are randomly generated
+func RootCAEqualStable(a, b *api.RootCA) bool {
+	switch {
+	case a == nil && b == nil:
+		return true
+	case a != nil && b != nil:
+	default:
+		return false
+	}
+	if !(bytes.Equal(a.CACert, b.CACert) &&
+		subtle.ConstantTimeCompare(a.CAKey, b.CAKey) == 1 &&
+		a.CACertHash == b.CACertHash) {
+		return false
+	}
+	switch {
+	case a.RootRotation == nil && b.RootRotation == nil:
+		return true
+	case a.RootRotation != nil && b.RootRotation != nil:
+	default:
+		return false
+	}
+	return (bytes.Equal(a.RootRotation.CACert, b.RootRotation.CACert) &&
+		subtle.ConstantTimeCompare(a.RootRotation.CAKey, b.RootRotation.CAKey) == 1 &&
+		bytes.Equal(a.RootRotation.CrossSignedCACert, b.RootRotation.CrossSignedCACert))
+}
+
+// ExternalCAsEqualStable compares lists of external CAs and determines whether they are equal
+func ExternalCAsEqualStable(a, b []*api.ExternalCA) bool {
```

Comment:
I think it would be easier to maintain that way. I can't see any scenario where we care about nil maps vs empty maps.

The same goes for `RootCAEqualStable`, except in that case you're using `subtle.ConstantTimeCompare`, so arguably there's a security reason to avoid using `reflect.DeepEqual` directly (though it's a weak one, and I'm not really sure it's worth the maintenance burden).